neuralhydrology/training/basetrainer.py

```diff
@@ -233,6 +262,15 @@ def train_and_validate(self):
                     print_msg += f" -- Median validation metrics: "
                     print_msg += ", ".join(f"{k}: {v:.5f}" for k, v in valid_metrics.items() if k != 'avg_total_loss')
                     LOGGER.info(print_msg)
+                
+
+                if self._early_stopping == True:
+                    if epoch > self._minimum_epochs_before_early_stopping:
+                        if (early_stopper.early_stop(valid_metrics['avg_total_loss'])):
+                            LOGGER.info(f"Early stopping triggered at epoch {epoch} with validation loss {valid_metrics['avg_total_loss']:.5f}. Training stopped.")
+                            break
+                if self._dynamic_learning_rate == True:
+                    scheduler.step(avg_losses['avg_total_loss'])
```

Comment:
The scheduler uses the _training_ loss to determine improvement (but the training loss rarely goes up).
I'd rather use `valid_metrics['avg_total_loss']` here as the metric (like for early stopping).